sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEUtils.scala

```diff
@@ -57,4 +58,20 @@ object AQEUtils {
       }
     case _ => Some(UnspecifiedDistribution)
   }
+
+  // Add an extra shuffle if input plan does not satisfy the required distribution.
+  // This method is invoked after optimizing skewed join in case we change final stage
+  // output partitioning.
+  def ensureRequiredDistribution(
+      plan: SparkPlan, distribution: Option[Distribution]): SparkPlan = distribution match {
+    case Some(d) if !plan.outputPartitioning.satisfies(d) =>
+      val numPartitions = d.requiredNumPartitions.getOrElse(conf.numShufflePartitions)
+      val shuffleOrigin = if (d.requiredNumPartitions.isDefined) {
+        REPARTITION_BY_NUM
+      } else {
+        REPARTITION_BY_COL
+      }
+      ShuffleExchangeExec(d.createPartitioning(numPartitions), plan, shuffleOrigin)
```

Comment:
I think the origin should always be `ENSIRE_REQUIREMENT`? The shuffle is added to satisfy the required distribution.

Reply:
We should retain origin shuffle here, otherwise some stage optimize rules  can break the `ENSIRE_REQUIREMENT` output partitioning in next stage. e.g. `CoalesceShufflePartitions` when the input plan's origin is `REPARTITION_BY_NUM`